plugin/entrySchema.go

```diff
@@ -0,0 +1,207 @@
+package plugin
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+type entrySchema struct {
+	TypeID    string         `json:"type_id"`
+	Label     string         `json:"label"`
+	Singleton bool           `json:"singleton"`
+	Actions   []string       `json:"actions"`
+	Children  []*EntrySchema `json:"children"`
+	entry     Entry
+}
+
+// EntrySchema represents an entry's schema. Use plugin.NewEntrySchema
+// to create instances of these objects.
+//
+// EntrySchema's a useful way to document your plugin's hierarchy. Users
+// can view your hierarchy via the stree command. For example, if you
+// invoke `stree docker` in a Wash shell (try it!), you should see something
+// like
+//
+// docker
+// ├── containers
+// │   └── [container]
+// │       ├── log
+// │       ├── metadata.json
+// │       └── fs
+// │           ├── [dir]
+// │           │   ├── [dir]
+// │           │   └── [file]
+// │           └── [file]
+// └── volumes
+//     └── [volume]
+//         ├── [dir]
+//         │   ├── [dir]
+//         │   └── [file]
+// 		└── [file]
+//
+// (Your output may differ depending on the state of the Wash project, but it
+// should be similarly structured).
+//
+// Every node must have a label. The "[]" are printed for non-singleton nodes;
+// they imply multiple instances of this thing. For example, "[container]" means
+// that there will be multiple "container" instances under the "containers" directory
+// ("container" is the label that was passed into NewEntrySchema). Similarly, "containers"
+// means that there will be only one "containers" directory (i.e. that "containers" is a
+// singleton). You can use EntrySchema#IsSingleton() to mark your entry as a singleton.
+//
+// TODO: Talk about how metadata schema's used to optimize `wash find` once that
+// is added.
+type EntrySchema struct {
+	// This pattern's a nice way of making JSON marshalling/unmarshalling
+	// easy without having to export these fields via the godocs. The latter
+	// is good because plugin authors should use the builders when setting them
+	// (so that we maintain a consistent API for e.g. metadata schemas).
+	//
+	// This pattern was obtained from https://stackoverflow.com/a/11129474
+	entrySchema
+}
+
+// NewEntrySchema returns a new EntrySchema object with the specified label.
```

Comment:
I think it'd be useful to document the way this is frequently used. i.e. `ChildSchema` usually invokes it with an empty instance of a child class, such as `NewEntrySchema(&childObj{}, "childObjLabel")`.

Reply:
I got it wrong too. Child implements
```
func (c *childObj) Schema() *plugin.EntrySchema {
  return plugin.NewEntrySchema(c, "childObjLabel")
}
```

Then parent uses it in ChildSchemas
```
(&childObj{}).Schema()
```